util/stop/stopper.go

```diff
@@ -0,0 +1,221 @@
+// and Stopper can manage these goroutines centrally to avoid leaks.
+// When Stopper's Stop method is called, if some tasks do not exit within the specified time,
+// the names of these tasks will be returned for analysis.
+type Stopper struct {
+	stopC   chan struct{}
+	cancels sync.Map // id -> cancelFunc
+	tasks   sync.Map // id -> name
+
+	atomic struct {
+		lastID    uint64
+		taskCount int64
+	}
+
+	mu struct {
+		sync.RWMutex
+		state state
+	}
+}
+
+// NewStopper create a stopper
+func NewStopper() *Stopper {
+	t := &Stopper{
+		stopC: make(chan struct{}),
+	}
+
+	t.mu.state = running
+	return t
+}
+
+// RunTask run a task that can be cancelled. ErrUnavailable returned if stopped is not running
+// See also `RunNamedTask`
+// Example:
+// err := s.RunTask(func(ctx context.Context) {
+// 	select {
+// 	case <-ctx.Done():
+// 	// cancelled
+// 	case <-time.After(time.Second):
+// 		// do something
+// 	}
+// })
+// if err != nil {
+// 	// hanle error
+// 	return
+// }
+func (s *Stopper) RunTask(task func(context.Context)) error {
+	return s.RunNamedTask("undefined", task)
+}
+
+// RunNamedTask run a task that can be cancelled. ErrUnavailable returned if stopped is not running
+// Example:
+// err := s.RunNamedTask("named task", func(ctx context.Context) {
+// 	select {
+// 	case <-ctx.Done():
+// 	// cancelled
+// 	case <-time.After(time.Second):
+// 		// do something
+// 	}
+// })
+// if err != nil {
+// 	// hanle error
+// 	return
+// }
+func (s *Stopper) RunNamedTask(name string, task func(context.Context)) error {
+	// we use read lock here for avoid race
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if s.mu.state != running {
+		return ErrUnavailable
+	}
+
+	id, ctx := s.allocate()
+	s.doRunCancelableTask(ctx, id, name, task)
+	return nil
+}
+
+// Stop stop all task in default timeout. If some tasks do not exit within the specified time,
+// the names of these tasks will be returned for analysis.
+func (s *Stopper) Stop() ([]string, error) {
+	return s.StopWithTimeout(defaultWaitStoppedTimeout)
+}
+
+// Stop stop all task in specified timeout. If some tasks do not exit within the specified time,
+// the names of these tasks will be returned for analysis.
+func (s *Stopper) StopWithTimeout(timeout time.Duration) ([]string, error) {
+	s.mu.Lock()
+	state := s.mu.state
+	s.mu.state = stopping
+	s.mu.Unlock()
+
+	switch state {
+	case stopped:
+		return nil, nil
+	case stopping:
+		<-s.stopC // wait concurrent stop completed
+		return s.runningTasks(), nil
+	}
+
+	defer func() {
+		close(s.stopC)
+	}()
+
+	s.cancels.Range(func(key, value interface{}) bool {
+		cancel := value.(context.CancelFunc)
+		cancel()
+		return true
+	})
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-timer.C:
+			return s.runningTasks(), errors.New("waitting for tasks complete timeout")
+		default:
+			if s.getTaskCount() == 0 {
+				return nil, nil
+			}
+		}
+
+		time.Sleep(time.Millisecond * 5)
```

Comment:
add comments to indicate that such 5ms delay can be a problem if we need to repeatedly create different stoppers, e.g. one stopper for each incoming request.